### PR TITLE
US99246/InitialedProfileImage

### PR DESCRIPTION
**Pull request by mtjandra**
More details can be found at http://design.d2l/components/profile-images/

* Given firstName, userId, and optionally lastName -> display initials
* colourId is required to have a consistent mapping to the background colour no default background coloured is specified in the spec
* Visually inspected in IE, Edge, Firefox and Chrome

image

### CODE REVIEW

File: d2l-profile-image.html [hunk 1]
```diff
@@ -8,21 +13,171 @@
...
+					display : flex;
+					align-items : center;
+					justify-content: center;
+				}
+				.shady-person {
+					--d2l-icon-fill-color: var(--d2l-color-mica);
+				}
+			</style>
+		</custom-style>
+		<a id="test"></a>
```
Comment (mtjandra) [inline d2l-profile-image.html]: Is there a cleaner way to handle these `dom-if` statements? I don't see any support for case statements or if/else-if/else

File: d2l-profile-image.html [hunk 2]
```diff
@@ -12,47 +13,154 @@
...
+					height: var(--size);
+					--d2l-icon-height: var(--size);
+					--d2l-icon-width: var(--size);
+					overflow: hidden;
+				}
+				:host([small]) {
+					display: inline-block;
+					--size: 30px;
+					border-radius: 4px;
+					@apply --d2l-label-text;
```
Comment (mtjandra) [inline d2l-profile-image.html]: Doesn't seem I can apply the same trick I did with `--size` when wanting to `@apply` something dynamically. The margin has to be specified afterwards each time because the text styling introducing some paddings/margins, order matters

File: d2l-profile-image.html [hunk 3]
```diff
@@ -12,47 +13,154 @@
...
+					display : flex;
+					align-items : center;
+					justify-content: center;
+				}
+				.shady-person {
+					--d2l-icon-fill-color: var(--d2l-color-ferrite);
+				}
+			</style>
+		</custom-style>
+		<template is="dom-if" if="[[_showShadyPerson(firstName, colourId)]]">
```
Comment (mtjandra) [inline d2l-profile-image.html]: Is there a cleaner way to handle these dom-if statements? I don't see any support for case statements or if/else-if/else

Comment (CarolBusman) [reply d2l-profile-image.html]: Don't know if it's cleaner, but maybe we could bind the classes on the two divs to computed properties that give either `shady-person` or `shady-person-hidden` (for example), and use css to show and hide the icon vs the initials, so the template would look something like:
```
<template>
	<div class="[[_shadyPersonClass]]"><d2l-icon icon="d2l-tier3:profile-pic"></d2l-icon></div>
	<div id="initials" class="[[_initialsClass]]">[[_getInitials(firstName, lastName)]]</div>
</template>
```
?

Comment (mtjandra) [reply d2l-profile-image.html]: Cool idea! Though I think I still prefer using `dom-if` in this situation. I find it better expresses the intent to hide/show elements just by reading the template, without having to check what class would be generated from the method. It also keeps the styling and logic separate. 

The reason I want to use case statements or if/elif/else is to guarantee exclusivity. The way it is now, I'm heavily relying on my tests.

Comment (svanherk) [reply d2l-profile-image.html]: Something like this:
```
		<template is="dom-if" if="[[_show('shady')]]">
			...
		</template>
		<template is="dom-if" if="[[_show('initials')]]">
			...
		</template>
...
properties: {
		_toDisplay: {
				type: String,
				value: 'shady'
		},
		...
...
function _show(string) {
  return _toDisplay == string;
}
```
And then run this every time attached or the href changes (once hypermedia):
```
if (should use shady) {
		_toDisplay = 'shady';
} else if (should use intitials) {
		_toDisplay = 'initials';
} else if ( should use image) {
		_toDisplay = 'image';
}
```

Comment (mtjandra) [reply d2l-profile-image.html]: Much closer to what I was looking for! I think I will look to refactor seperatley

**Comment by mtjandra** [comment]
@CarolBusman @svanherk @jeroach any concerns that need to be addressed?

File: d2l-profile-image.html [hunk 4]
```diff
@@ -12,47 +13,154 @@
...
+					--size: 30px;
+					border-radius: 4px;
+					@apply --d2l-label-text;
+					margin: 0;
+				}
+				:host([medium]) {
+					display: inline-block;
+					--size: 42px;
+					border-radius: 6px;
+					@apply --d2l-heading-4;
```
Comment (svanherk) [inline d2l-profile-image.html]: The typography mixins to apply all match the spec, but the `font-weight`s seem off.  In the images on design.d2l, it looks like extra-large has a `font-weight` of `700` and the rest are `400`, but here it looks like the reverse is happening:
image
I think we'll want to ask Bob if the `font-weight`s should be different than the default for those.

Comment (mtjandra) [reply d2l-profile-image.html]: @thehappypixel

File: d2l-profile-image.html [hunk 5]
```diff
@@ -12,47 +13,154 @@
...
+						'#5F727D',
+						'#3B4148',
+						'#59616C'
+					],
+					readOnly: true
+				}
+			},
+			attached: function() {
+				this.async(function() {
+					var initials = Polymer.dom(this.root).querySelector('#initials');
```
Comment (svanherk) [inline d2l-profile-image.html]: Rather than check if the element exists and set the attribute, I think you want to do this:
`this.updateStyles({'--initials-background-color': backgroundColour});`

And in the `.initials` css class above, have this:
`background-color: var(--initials-background-color);`

File: d2l-profile-image.html [hunk 6]
```diff
@@ -12,47 +13,154 @@
...
+						'#D13B7F',
+						'#47565E',
+						'#5F727D',
+						'#3B4148',
+						'#59616C'
+					],
+					readOnly: true
+				}
+			},
+			attached: function() {
```
Comment (svanherk) [inline d2l-profile-image.html]: This will only run once, when the component is attached, so changing the `colour-id` attribute doesn't cause an update the same way changing `first-name` and `last-name` does.  Maybe this is fine, but what you could do instead is add an observer that will run whenever that attribute changes:
```
			properties: {
				...
				colourId: {
					type: Number,
					value: -1,
					observer: '_handleColourId'
				},
				...
			_handleColourId: function(newColourId) {
				// map newColourId to backgroundColour
				this.updateStyles({'--initials-background-color': backgroundColour});
			},
```

File: test/d2l-profile-image.html [hunk 7]
```diff
@@ -56,6 +63,52 @@
 					});
 				});
 
+				[
+					{ firstName:'first', lastName: 'last', expectedInitials: 'FL', colourId: 10, valid: true },
```
Comment (svanherk) [inline test/d2l-profile-image.html]: Nice coverage, I was gonna ask about the first name / last name edge cases but they're all outlined here!

File: d2l-profile-image.html (hunk 5 above)
Comment (mtjandra) [reply d2l-profile-image.html]: Looks like it is working well

Comment (mtjandra) [reply d2l-profile-image.html]: Think the problem was trying to use constants. I thought the default values for properties would always be ready whenever I need them. Turns out this is not the case. So no way to provide constants?

**Comment by ableeck** [comment]
If I understand this implementation correctly, then - no matter what the privacy settings are - the browser is seeing the first name and last name for a user. Hence it would be reasonably easy for an end-user to find out.
Should the decision of what information to provide not be made at the API level, so that only the part of the personal information that is supposed to be shown gets sent to the browser?

**Comment by mtjandra** [comment]
> Should the decision of what information to provide not be made at the API level, so that only the part of the personal information that is supposed to be shown gets sent to the browser?

Yes that's correct the API is responsible for responding with the correct information not the web-component.

**Comment by ableeck** [comment]
> Yes that's correct the API is responsible for responding with the correct information not the web-component.

Thanks, Miles!

**Comment by mtjandra** [comment]
All comments addressed. I can follow up with bob about the font size separate to confirm if he wanted to deviate from the default styling of the text specified in the spec.

**Comment by mtjandra** [comment]
@svanherk I checked in with Bob, we will leave the font-weights as is

File: d2l-profile-image.html (hunk 4 above)
Comment (thehappypixel) [reply d2l-profile-image.html]: Nice catch @svanherk - I made a mistake in my documentation and should not have deviated from the typography styles. I spoke with Miles on slack and I'd like to maintain consistency with our typography mixins. I will update the image on design.d2l. Thanks!!

File: src/d2l-profile-image-shared-styles.html [hunk 8]
```diff
@@ -1,40 +1,51 @@
...
-		<style>
-			:host {
-				display: none;
-				width: var(--size);
-				height: var(--size);
-				--d2l-icon-height: var(--size);
-				--d2l-icon-width: var(--size);
-				overflow: hidden;
-			}
+		<custom-style include="d2l-typography">
```
Comment (svanherk) [inline src/d2l-profile-image-shared-styles.html]: Sent details offline about this use of `custom-style`, we can clean this up in another PR

Comment (mtjandra) [reply src/d2l-profile-image-shared-styles.html]: Updated

**Comment by mtjandra** [comment]
@svanherk @CarolBusman any other concerns?

File: d2l-profile-image.html [hunk 9]
```diff
@@ -13,25 +13,119 @@
...
+					--d2l-icon-fill-color: var(--d2l-color-ferrite);
+				}
+				@media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
+					.shady-person{
+							width: 100%;
+							height: 100%;
+					}
+				}
+			</style>
+		</custom-style>
```
Comment (svanherk) [inline d2l-profile-image.html]: Formatting is off here, I think you want to remove this line and un-indent the section above to align the `style` tags.

File: src/d2l-profile-image-shared-styles.html [hunk 10]
```diff
@@ -1,8 +1,9 @@
 <link rel="import" href="../../polymer/polymer.html">
+<link rel="import" href="../../d2l-typography/d2l-typography.html">
 
 <dom-module id="d2l-profile-image-shared-styles">
 	<template>
-		<style>
+	<style is="custom-style" include="d2l-typography">
```
Comment (svanherk) [inline src/d2l-profile-image-shared-styles.html]: Minor: I think you want to indent this line once more

**Comment by mtjandra** [comment]
> Some formatting changes above. I also made #10 to document some inconsistencies I'm seeing, but don't want to block this PR on those.

Thanks I will format them at a later point
